### ISSUE TITLE
(Suggestion) Make SHA-512 be the one that's tested first when autodetection is selection.

### ISSUE THREAD

**Issue by nikkopt** [veracrypt/VeraCrypt#792]
Greetings. 
So, when i mount a volume and i manually select the PKCS-5 (SHA512 because it's the default when creating volumes), the volume mount's in like 1.5 seconds. When "Autodetect" is selected, it takes like 15 seconds to mount which makes me think that the default one for creating volumes isn't the first one to be tested on the list.
Is there a good reason for this 15 seconds wait time? Can't SHA512 be the first to be tested since it's de default one? 
Thanks.

**Comment by alt3r-3go** [comment]
There's a feature for that, under Default Mount options (see here: https://www.veracrypt.fr/en/Default%20Mount%20Parameters.html) - you can set the default PRF to try first right there. Does that help?

**Comment by nikkopt** [comment]
It does.
I feel dumb missing that. Thanks

**Comment by cypherbits** [comment]
Setting the Default Mount Parameters and making SHA-512 to be the first tested when Autodetection is selected is a different thing. I think this should be reopened. 
When you first create a volume, the default selected option is SHA-512, so if you select Autodetection here, the first checked should be in the same order.

**Comment by nikkopt** [comment]
I agree, although i think setting it in the options is a good workaround. I'll reopen it anyway because it does make sense.

**Comment by alt3r-3go** [comment]
Looks like this is going to be fixed with https://github.com/veracrypt/VeraCrypt/commit/fdf7888ab3675a267e06e2f3acceeedfa5b74f62

**Comment by idrassi** [comment]
I confirm that a change was implemented to make PRF auto-detection nearly fast as manual selection of the PRF. The newer commit https://github.com/veracrypt/VeraCrypt/commit/4b98ff0e9810a218f802d08cfd546c2fd67757dc brings better code than previous commit mentioned by @alt3r-3go.

I have uploaded the first beta of 1.24-Update9 containing this enhancement to Sourceforge [Nightly Build folder ](https://sourceforge.net/projects/veracrypt/files/VeraCrypt%20Nightly%20Builds/)(look for VeraCrypt Setup 1.24-Update9-Beta-21-08-07.exe or VeraCrypt Portable 1.24-Update9-Beta-21-08-07.exe). Binaries are signed with official keys as usual.

Tests and feedback are welcomed.

**Comment by nikkopt** [comment]
Yes both are using the same encryption parameters. 
The NTFS containers only take 13-15 seconds if i choose auto, they mount in 1-2 seconds if the correct option is chosen.
Windows event viewer doesn't help much, it just says `Volume Z: (\Device\VeraCryptVolumeZ) está em bom estado de funcionamento. Não é necessária nenhuma ação.` which means that volume is healthy and no action is needed.
I don't know if this helps but when veracrypt is mounting those volumes, the kernel spawns 3 veracrypt processes (each using one thread at 100%) as you can see here: https://i.imgur.com/vXW1nDC.png . No disk access.

**Comment by idrassi** [comment]
Thank you for the detailed analysis.

The 3 threads that use 100% CPU each are part of the threads used by VeraCrypt to try all possible hash algorithms in parallel (SHA512, SHA256, Whirpool, RIPEMD160 and Streebog). 

I guess you took this screenshot after 2 seconds of the beginning of mount. In this case, the 3 threads correspond to the slowest hashes (Whirpool, RIPEMD160 and Streebog) and normally SHA512 and SHA256 have already finished which means that mounting should have returned.

How many virtual processors does your CPU have?
Can you please post screenshot of VeraCrypt benchmark of PKCS5-PRF (menu Tools->Benchmark then choose PKCS5-PRF)?
Did you limit the number of threads to use in VeraCrypt settings?

I hope this will give me some idea on why NTFS is affected by this.

Last query: do you have the same delay if you mount the NTFS volumes as removal media (Mount Options -> Mount volume as removable medium)?

Thanks.

**Comment by nikkopt** [comment]
>used by VeraCrypt to try all possible hash algorithms in parallel

You're right, it starts with 5. It just seems to mount when all of them end. 

https://user-images.githubusercontent.com/2060960/153775434-62dc1dad-b583-4ba9-8114-86c43bc98212.mp4

> How many virtual processors does your CPU have?
16. Dual threaded 8 cores.

> Can you please post screenshot of VeraCrypt benchmark of PKCS5-PRF (menu Tools->Benchmark then choose PKCS5-PRF)?
![pkcs](https://user-images.githubusercontent.com/2060960/153774702-f478fe24-2b31-4370-a4e0-3f8829807651.PNG)

> Did you limit the number of threads to use in VeraCrypt settings?

Just checked afinnity, it's using all of them. For fun i tried with just 4 and it was the same. I may be wrong but i don't think it matters. When it's mounting, the veracrypt.exe process isn't using any cpu resources, in the windows task manager it's a process called "System". Using the app "TaskExplorer" (the one i screenshot in my last post), i can actually see those veracrypt.sys child processes that were spawned by the kernel.

> do you have the same delay if you mount the NTFS volumes as removal media (Mount Options -> Mount volume as removable medium)?

Yep, same thing

**Comment by idrassi** [comment]
Thank you very much for the video and the information.
The thread pool mechanism works as expected but somehow the notification that SHA512 has finished is not received by the mounting thread. 
I'm still scratching my head on why this happens and why only with NTFS volumes.
I will review the source code in case I can find something. I will keep you informed.

**Comment by nikkopt** [comment]
Thanks a bunch.
By the way, i only tested with FAT and NTFS, not the other filesystems. I'll try them tomorrow since i'm not at the computer at the moment.

**Comment by nikkopt** [comment]
It seems it only affects NTFS (I can't try ReFS, it gives me an error saying windows couldn't format it). No filesystem, FAT, FAT32 and exFAT mount instantaneously.
If i mount the NTFS container and then format it to another filesystem through the explorer, it will begin mounting instantaneously. If i format it again to NTFS, it will take again 13-15 seconds to mount. 
This makes one think that it's a windows problem. However, wouldn't it also behave like that if i chose SHA512 manually? It doesn't. It mounts instantaneously if the hash is chosen.
Is there anything else i can try to help troubleshoot the problem?

PS: For the containers that mount instantaneously, the threads trying the other hashes keep going and only stop when they fail, not when the right one is found and the container is mounter. This is by design?

**Comment by nikkopt** [comment]
I did some more tests.

**On a windows 7 virtual machine with 4 cores:**
Manual:~1 sec
Auto:~10 secs (yep, it's a bit faster on a VM with less cores than on the host machine)

**On an old dual core celeron netbook running a very old version of win10:**
Manual:~4 secs
Auto:~18secs

These machines have nothing in common other than being mine.
By the way, i can greatly decrease the time it takes to mount on my main machine from 15secs to around 6 by changing the power profile (which locks the cpu at a max 2.2GHz) from power saving to balanced or performance. But by doing that, the 1 second it takes to mount a volume by manually choosing SHA512 drops literally to instantaneously. What i'm trying to say is that the difference between auto and manual in percentage stays the same.

**Comment by idrassi** [comment]
Thank you for all these tests, I appreciate your help.
The Windows 7 VM test is just incredible! But this can be caused by Windows 7 being more optimal than Windows 10 in such use case.

As you noted, it is strange that the issue with NTFS disappears when using manual PRF selection. It seems as if Windows is doing something special for NTFS volume and it is somehow blocking VeraCrypt main thread until all its sub threads finish their processing.

I will try to reproduce the issue on my various machines. If I can, then I will be able to generate driver traces and from there look for the culprit.

Concerning your question about the other threads continuing their processing even if the right hash was found, this is indeed by design since there is no way to stop these threads in the middle of the job and so we just leave them until they fail.

**Comment by eugenesvk** [comment]
I've just checked on an old Macbook OSX 10.15 Veracrypt v1.25.9 (with macfuse 4.4.1) with a small 50Mb container with both hidden and outer volumes formatted as **FAT**, and I observe exactly the same issue - the hidden volume autodetection is ~20 seconds while selecting SHA-512 manually takes ~2 seconds
The outer volume doesn't have this difference, both autodetection and manual detection take ~2sec
(both volumes are encrypted with a single-digit password)

Sorry don't know how to do any proper tests on a Mac, so just a couple observations based off activity monitor
I see two procs here, one with ~10 threads, another one with 1
While trying to mount:
 - the ~10-thread proc increases the number by a few threads in both instances,
 - the 1-thread proc behaves differently, but the only difference is in how % of CPU and time it takes (autodetection is much longer)

**Comment by Progress1** [comment]
- creating a volume - the default selected option is SHA-512
- i set default mount parameters in setting: SHA-512, then manual mount via GUI - mount is quick
- using command line (without /hash parameter, using autodetection) - mount is slow !!!!
- using command line (with /hash sha512 parameter) - mount is quick

so it looks that SHA-512 it's still NOT checked as first. Using last version 1.26.24.

Using hash type in command line leaks information about encryption!

**Comment by idrassi** [comment]
Thanks for the additional reports.

There are two different topics mixed in this issue.

First, SHA-512 is already the first KDF tried by autodetection. On Windows, PRF autodetection has also used parallel derivation since 1.24-Update9, so the delay seen in 1.26.24 is not caused by SHA-512 being tried last.

The remaining delay in 1.26.24 is due to the other KDF derivations continuing after a successful match in some paths. This has been
improved after 1.26.24 by adding an abort mechanism for PBKDF2/Argon2 workers on Windows (9dc24ba7). Linux/macOS also received parallel header KDF autodetection (`a173a11c`). These changes are included in the 1.26.28 beta builds, so testing with [1.26.28 Beta4](https://github.com/veracrypt/VeraCrypt/releases/tag/VeraCrypt_1.26.28_Beta4) would be useful.

For the command-line case: when `/hash` is omitted, VeraCrypt tries all possible PRFs. Also, plain `/q` does not load user preferences. If you want the Default Mount Parameters value, such as SHA-512, to be used without putting `/hash sha512` on the command line, use:

`/q preferences`

This avoids exposing the KDF in the command line, although the preference is stored in VeraCrypt's configuration and applies globally.

If [1.26.28 Beta4](https://github.com/veracrypt/VeraCrypt/releases/tag/VeraCrypt_1.26.28_Beta4) is still slow, please post:
- the exact command line with password/keyfile paths removed
- whether `/q` or `/q preferences` was used
- whether the volume is normal or hidden
- filesystem inside the volume
- timings for autodetect, `/q preferences`, and explicit `/hash sha512`

For hidden volumes, autodetection can still be slower than manual SHA-512 because VeraCrypt must first rule out the normal header before trying the hidden header. Improving that further would require a deeper change to the autodetection ordering across header locations.

**Comment by idrassi** [comment]
Closing this as done in 1.26.29.
